executor/new_distsql.go

```diff
@@ -189,10 +201,202 @@ func (e *IndexReaderExecutor) doRequest() error {
+// Then we hold the returning rows and finish this task.
+func (e *IndexLookUpExecutor) executeTask(task *lookupTableTask) {
+	var err error
+	defer func() {
+		task.doneCh <- err
+	}()
+	tableReader := &TableReaderExecutor{
+		asName:  e.asName,
+		table:   e.table,
+		tableID: e.tableID,
+		dagPB:   e.tableRequest,
+		schema:  e.schema,
+		ctx:     e.ctx,
+	}
+	err = tableReader.doRequestForHandles(task.handles)
+	if err != nil {
+		return
+	}
+	for {
+		var row *Row
+		row, err = tableReader.Next()
+		if err != nil || row == nil {
+			return
+		}
+		task.rows = append(task.rows, row)
+	}
+}
+
+// fetchHandles fetches a batch of handles from index data and builds the index lookup tasks.
+// We initialize some workers to execute this tasks concurrently and put the task to taskCh by order.
+func (e *IndexLookUpExecutor) fetchHandles() {
+	lookupConcurrencyLimit := e.ctx.GetSessionVars().IndexLookupConcurrency
+	workCh := make(chan *lookupTableTask, lookupConcurrencyLimit)
+	defer func() {
+		close(e.taskChan)
+		close(workCh)
+		e.result.Close()
+	}()
+
+	for i := 0; i < lookupConcurrencyLimit; i++ {
+		go func() {
+			childCtx, cancel := goctx.WithCancel(e.ctx.GoCtx())
+			defer cancel()
+			select {
+			case task := <-workCh:
+				if task == nil {
+					return
+				}
+				e.executeTask(task)
+			case <-childCtx.Done():
+			}
+		}()
+	}
+
+	txnCtx := e.ctx.GoCtx()
+	for {
+		handles, finish, err := extractHandlesFromIndexResult(e.result)
+		if err != nil || finish {
+			e.tasksErr = errors.Trace(err)
+			return
+		}
+		tasks := e.buildTableTasks(handles)
+		for _, task := range tasks {
+			select {
+			case <-txnCtx.Done():
+				return
+			case workCh <- task:
+			}
+			e.taskChan <- task
+		}
+	}
+}
+
+func (e *IndexLookUpExecutor) buildTableTasks(handles []int64) []*lookupTableTask {
+	// Build tasks with increasing batch size.
+	var taskSizes []int
+	total := len(handles)
+	batchSize := e.ctx.GetSessionVars().IndexLookupSize
+	for total > 0 {
+		if batchSize > total {
+			batchSize = total
+		}
+		taskSizes = append(taskSizes, batchSize)
+		total -= batchSize
+	}
+
+	var indexOrder map[int64]int
+	if e.keepOrder {
+		// Save the index order.
+		indexOrder = make(map[int64]int, len(handles))
+		for i, h := range handles {
+			indexOrder[h] = i
+		}
+	}
+
+	tasks := make([]*lookupTableTask, len(taskSizes))
+	for i, size := range taskSizes {
+		task := &lookupTableTask{
+			handles:    handles[:size],
+			indexOrder: indexOrder,
+		}
+		task.doneCh = make(chan error, 1)
+		handles = handles[size:]
+		tasks[i] = task
+	}
+	return tasks
+}
+
+// Schema implements Exec Schema interface.
+func (e *IndexLookUpExecutor) Schema() *expression.Schema {
+	return e.schema
+}
+
+// Close implements Exec Close interface.
+func (e *IndexLookUpExecutor) Close() error {
+	e.result = nil
+
+	// Consume the task channel in case channel is full.
+	for range e.taskChan {
```

Comment:
We hope the result stop, but it will still fetch exhaustively. But this case is not always happened, for SQL like `select * from t where a = 1 limit 1`, we should push limit 1 to index plan, so even if there are 1000 rows of a = 1, we will still fetch 1 handle.